src/test/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtilsTest.java

```diff
@@ -108,11 +116,72 @@ public InputStream answer(InvocationOnMock invocation) throws Throwable {
             httpClient,
             requestBuilder);
 
-    Mockito.verify(requestBuilder, Mockito.times(4))
+    Mockito.verify(requestBuilder, Mockito.times(1))
         .generateStreamingIngestPostRequest(Mockito.anyString(), Mockito.any(), Mockito.any());
     Assert.assertEquals("honk", result.getMessage());
   }
 
+  @Test
+  public void testJWTRetries() throws Exception {
+    SnowflakeURL url = new SnowflakeURL(TestUtils.getAccountURL());
+    CloseableHttpClient httpClient = HttpUtil.getHttpClient(TestUtils.getAccount());
+
+    JWTManager manager =
+        new JWTManager(url.getAccount(), TestUtils.getUser(), TestUtils.getKeyPair(), null);
+    JWTManager spyManager = Mockito.spy(manager);
+
+    // inject spy manager
+    RequestBuilder requestBuilder =
+        new RequestBuilder(
+            url.getAccount(),
+            TestUtils.getUser(),
+            TestUtils.getKeyPair(),
+            url.getScheme(),
+            TestUtils.getHost(),
+            url.getPort(),
+            "testJWTRetries",
+            spyManager,
+            httpClient,
+            "testJWTRetries");
+
+    // build payload
+    Map<Object, Object> payload = new HashMap<>();
+    if (!TestUtils.getRole().isEmpty() && !TestUtils.getRole().equals("DEFAULT_ROLE")) {
+      payload.put("role", TestUtils.getRole());
+    }
+    ObjectMapper mapper = new ObjectMapper();
+
+    // request wih invalid token
+    doReturn("invalid_token").when(spyManager).getToken();
+    try {
+      ChannelsStatusResponse response =
+          executeWithRetries(
+              ChannelsStatusResponse.class,
+              CLIENT_CONFIGURE_ENDPOINT,
+              mapper.writeValueAsString(payload),
+              "client configure",
+              STREAMING_CLIENT_CONFIGURE,
+              httpClient,
+              requestBuilder);
+      Assert.fail("Expected error for invalid token");
+    } catch (SecurityException ignored) {
+    }
+
+    // request with invalid token first time but with valid token second time after refresh
+    doReturn("invalid_token").doReturn(manager.getToken()).when(spyManager).getToken();
+    ChannelsStatusResponse response =
+        executeWithRetries(
+            ChannelsStatusResponse.class,
+            CLIENT_CONFIGURE_ENDPOINT,
+            mapper.writeValueAsString(payload),
+            "client configure",
+            STREAMING_CLIENT_CONFIGURE,
+            httpClient,
+            requestBuilder);
+
+    assert (response.getStatusCode() == RESPONSE_SUCCESS);
```

Comment:
can you also add a test which makes sure it doesnt go in an infinite loop? for instance max it should retry is 3 and after that it should go in securityException

Reply:
Added [here](https://github.com/snowflakedb/snowflake-ingest-java/blob/6927aaaf0b68cfa6c1e40df28512a535d0f6a661/src/test/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtilsTest.java#L164) and [here](https://github.com/snowflakedb/snowflake-ingest-java/blob/6927aaaf0b68cfa6c1e40df28512a535d0f6a661/src/test/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtilsTest.java#L184).